現段階でのコードレビューをお願いします。

kitakkun opened the pull request.

kitakkun committed:
create navigation from HomeScreen to ImageDetailScreen

kitakkun committed:
refactor MainScreen

kitakkun committed:
fix a problem that bottom bar remains on ImageDetailScreen or PostDetailScreen

kitakkun committed:
organize packages

kitakkun committed:
remove unused code in PostViewModel

kitakkun committed:
move FileUtils to util package

kitakkun committed:
remove fake post lists

kitakkun committed:
renamed FirebaseMediator to FirebaseAuthManager

kitakkun committed:
Merge pull request #2 from kitakkun/feature/refactor

[Feature/refactor] finish refactoring code

kitakkun committed:
implement list items' snapping when scrolling

kitakkun committed:
Merge pull request #3 from kitakkun/feature/enable-snapping-on-imagedetailscreen

Feature/enable snapping on ImageDetailScreen

kitakkun committed:
refactor ImageDetailScreen

kitakkun committed:
complete experimental UI of PostDetailScreen

kitakkun committed:
add reaction button

kitakkun committed:
add reaction option popup, add snaps on AttachedImagesDisplay, emoji support

kitakkun committed:
reduce font size of userId text

kitakkun committed:
setup onClick functions

kitakkun committed:
Merge pull request #4 from kitakkun/feature/create-post-detail-ui

Feature/create post detail UI

kitakkun committed:
implement Reaction data model and its Repository class

kitakkun committed:
Merge pull request #6 from kitakkun/feature/create-reaction-backend

implement Reaction data model and its Repository class

kitakkun committed:
add reaction-registering flow

kitakkun committed:
create a base of UserProfileScreen

kitakkun committed:
create a base of UserProfileScreen

kitakkun committed:
create Follow data class

kitakkun committed:
complete follow system backend

kitakkun committed:
Merge pull request #7 from kitakkun/feature/follow-system-backend

Feature/follow system backend

kitakkun committed:
add ServerTimestamp annotation and write some comments

kitakkun committed:
refactor PostsRepository

kitakkun committed:
refactor UsersRepository

kitakkun committed:
refactor PostsRepository and ReactionRepository

kitakkun committed:
remove unnecessary class

kitakkun committed:
rename data class of database to more explicit one

kitakkun committed:
Merge pull request #8 from kitakkun/feature/refactor-data-layers

Feature/refactor data layers

kitakkun committed:
Merge branch 'feature/userprofilescreen' of https://github.com/kitakkun/FooS into feature/userprofilescreen

kitakkun committed:
create GetReactionsByUserIdUseCase

kitakkun committed:
update code related to ReactionScreen

kitakkun committed:
implement userProfile screen

kitakkun committed:
Merge pull request #10 from kitakkun/feature/userprofilescreen

Feature/userprofilescreen

kitakkun committed:
enable image cache (but slower performance on the first load)

kitakkun committed:
fix a bug on repositories

- fix an error on PostsRepository.create (document reference error)
- fix reactions query was not working properly on ReactionsRepository
- implement GetReactionsByUserIdUseCase

kitakkun committed:
update reaction data structure, fix printed reactions are by user itself, not received ones.

kitakkun committed:
improve ReactionScreenUI a bit

kitakkun committed:
Merge branch 'develop' into feature/reaction-screen

kitakkun committed:
Merge pull request #11 from kitakkun/feature/reaction-screen

Feature/reaction screen

kitakkun committed:
fix duplication of the definition of provideReactionsRepository

kitakkun committed:
fix error on onReactionButtonClicked

kitakkun committed:
Merge pull request #13 from kitakkun/feature/process-reaction-event

Feature/process reaction event

kitakkun committed:
create ImageConverter

kitakkun committed:
enable auto compress feature when uploading images

when creating a post, attached images are automatically converted to the specified size.

kitakkun committed:
extract max pixel size as a private const value(updated to 1024)

kitakkun committed:
Merge pull request #14 from kitakkun/feature/auto-compress-images-before-send

Feature/auto compress images before send

kitakkun committed:
add comment and optimize imports

kitakkun committed:
remove PreloadAsyncImage because it produces too much requests

kitakkun committed:
Merge pull request #15 from kitakkun/feature/image-caching

Feature/image caching

kitakkun committed:
update post content ui

add splitter between two postItems and add elapsed time or date text after the userId.

kitakkun committed:
Merge pull request #16 from kitakkun/feature/improve-timeline-ui

update post content ui

kitakkun committed:
implement initial image index on ImageDetailView

kitakkun committed:
Merge pull request #17 from kitakkun/feature/ImageDetailScreen-by-selected-image

implement initial image index on ImageDetailView

kitakkun committed:
implement navigation from UserProfile to PostDetail

kitakkun committed:
rewrite HomeScreen navigations using SharedFlow

kitakkun committed:
Merge pull request #18 from kitakkun/feature/fix-home-screen-navigation-flow

rewrite HomeScreen navigations using SharedFlow

kitakkun committed:
extract PostList as a common composable

kitakkun committed:
Merge pull request #19 from kitakkun/feature/extract-postlist-to-common-composable

extract PostList as a common composable

kitakkun committed:
split nav code into the parts

kitakkun committed:
place bottomNavBar visibility logic into ScreenBottomNavBar Composable

kitakkun committed:
Merge pull request #20 from kitakkun/feature/refactor-bottom-nav-bar

Feature/refactor bottom nav bar

kitakkun committed:
implement ScreenViewModel (shared viewModel)

kitakkun committed:
Merge pull request #21 from kitakkun/feature/bottom-nav-event

implement ScreenViewModel (shared viewModel)

kitakkun committed:
remove posting progress screen

kitakkun committed:
rewrite navigate up event by SharedFlow

kitakkun committed:
fix bottom nav event not working properly

forgot to remove test code

kitakkun committed:
Merge pull request #22 from kitakkun/feature/bottom-nav-event

fix bottom nav event not working properly

kitakkun committed:
Merge pull request #23 from kitakkun/feature/async-post

Feature/async post

kitakkun committed:
change order of posts to descending

kitakkun committed:
implement auto fetching older posts at HomeScreen

kitakkun committed:
update repositories

add fetchByUserIdWithDate on PostRepository, rename fetchPost to fetchByPostId, fetchPostsByUserId to fetchByUserId...

kitakkun committed:
rename constant MAX_LOAD_COUNT to DEFAULT_LOAD_LIMIT

kitakkun committed:
update repositories and fix wrong code in UsersRepository

kitakkun committed:
Merge pull request #24 from kitakkun/feature/data-layer

Feature/data layer

kitakkun committed:
remove unnecessary usecase

kitakkun committed:
refactor HomeScreen

kitakkun committed:
implement auto fetch feature on UserProfileScreen

kitakkun committed:
Merge pull request #25 from kitakkun/feature/auto-fetch-older-posts

Feature/auto fetch older posts

kitakkun committed:
implement scroll to top when selecting home menu at HomeScreen

kitakkun committed:
Merge pull request #26 from kitakkun/feature/scroll-up-at-homescreen

implement scroll to top when selecting home menu at HomeScreen

kitakkun committed:
implement onContentClick and onImageClick Behaviors on UserProfileScreen

kitakkun committed:
Merge branch 'develop' into feature/onclick-handling-at-userprofile

kitakkun committed:
implement onUserIconClick at UserProfileScreen

kitakkun committed:
Merge pull request #27 from kitakkun/feature/onclick-handling-at-userprofile

Feature/onclick handling at userprofile

kitakkun committed:
reformat code

kitakkun committed:
add top bar when needed

kitakkun committed:
implement follow button on UserProfileScreen

kitakkun committed:
fix a problem that users can follow themselves

kitakkun committed:
extract string resources

kitakkun committed:
Merge pull request #28 from kitakkun/feature/follow

Feature/follow

kitakkun committed:
recreate LazyColumn in UserProfileScreen

kitakkun committed:
Merge pull request #29 from kitakkun/feature/fix-user-profile-remains-top-userprofilescreen

recreate LazyColumn in UserProfileScreen

kitakkun committed:
some stuff (just push the progress)

kitakkun committed:
implement nested vertical scroll with TabRow

need to use jetpack compose alpha02 or above. In the previous version, there is a bug on velocity of scrolling.

kitakkun committed:
Merge pull request #30 from kitakkun/feature/tab-swiping

Feature/tab swiping

kitakkun committed:
remove centering of contents when the amount of row items is small

kitakkun committed:
replace posts tab with PostItemList, adjust horizontal scroll area on HorizontalPager

kitakkun committed:
fix wrong following and followers counts

kitakkun committed:
Merge pull request #31 from kitakkun/feature/refactor-user-profile-screen

Feature/refactor user profile screen

kitakkun committed:
replace with VerticalNestedScrollView

kitakkun committed:
Merge pull request #32 from kitakkun/feature/refactor-user-profile-screen

replace with VerticalNestedScrollView

kitakkun committed:
create followerList UI and ViewModel (in progress)

kitakkun committed:
fix setting screen icon border

kitakkun committed:
extract UserIdentity texts as a common component, fix collapsed 'following' button

kitakkun committed:
extract follow button as a shared component

kitakkun committed:
add following and followers tabs

kitakkun committed:
implement auto fetch followers and followees, hide follow button in a specific condition

kitakkun committed:
make button outlined when following

kitakkun committed:
add initial page branch by "following" or "followers"

kitakkun committed:
add comment, some refactor

kitakkun committed:
create navigation to UserProfile

kitakkun committed:
fix a bug that userItem multiplies per resuming

kitakkun committed:
Merge branch 'develop' into feature/follow-list-screen

kitakkun committed:
Merge pull request #33 from kitakkun/feature/follow-list-screen

Feature/follow list screen

kitakkun committed:
fix a problem that navigation from UserProfile to FollowList was not working

kitakkun committed:
implement location select feature

kitakkun committed:
implement location confirmation screen

kitakkun committed:
replace StateFlow code with State

kitakkun committed:
Merge pull request #34 from kitakkun/feature/stateflow-to-state

replace StateFlow code with State

kitakkun committed:
fix followList fetch timing

kitakkun committed:
Merge pull request #35 from kitakkun/feature/fix-follow-list-fetch-timing

fix followList fetch timing

kitakkun committed:
implement location selection feature

kitakkun committed:
add permission check on the bottom toolbar

kitakkun committed:
add initial camera position control to GoogleMap

kitakkun committed:
Merge pull request #36 from kitakkun/feature/location-selector

Feature/location selector

kitakkun committed:
implement simple bubble on map

kitakkun committed:
Merge pull request #37 from kitakkun/feature/map-overlay-bubble

implement simple bubble on map

kitakkun committed:
Merge pull request #38 from kitakkun/feature/top-app-bar

add top bar when needed

kitakkun committed:
refactor Screens definition

kitakkun committed:
extract UserItem composable

kitakkun committed:
change the way to resolve navArguments.

kitakkun committed:
move PostItem and extract ReactionItem

kitakkun committed:
refactor code along navigation system, and padding values

kitakkun committed:
extract arguments' keys, fix FollowList navigation logic

kitakkun committed:
extract Composables in UserProfileScreen

kitakkun committed:
adjust username and userId UI, delete unnecessary files

kitakkun committed:
reformat code replace literals with constants

kitakkun committed:
extract list composables

kitakkun committed:
Merge pull request #39 from kitakkun/feature/refactor-whole-code

Feature/refactor whole code

kitakkun committed:
extract PostDetailView

kitakkun committed:
implement postDetail bottom sheet

kitakkun committed:
fix unexpected re-composition

kitakkun committed:
Merge pull request #40 from kitakkun/feature/bottom-sheet-on-mapscreen

Feature/bottom sheet on mapscreen

kitakkun committed:
Implement reaction deletion, add visual indicator of user's reaction

kitakkun committed:
implement media view on profile screen

kitakkun committed:
Merge pull request #41 from kitakkun/feature/media-view-on-profile

implement media view on profile screen

kitakkun committed:
Merge pull request #42 from kitakkun/feature/reaction

Implement reaction deletion, add visual indicator of user's reaction

kitakkun committed:
refactor useCases (in progress)

kitakkun committed:
Update MapViewModel.kt

kitakkun committed:
refactor useCases and data classes

kitakkun committed:
implement fetching on UserProfileScreen

kitakkun committed:
Merge pull request #44 from kitakkun/feature/unify-data-objects

Feature/unify data objects

kitakkun committed:
fix a problem that userReactedPosts was ordered randomly

kitakkun committed:
Merge pull request #45 from kitakkun/feature/fix-reacted-post-random-order

fix a problem that userReactedPosts was ordered randomly

kitakkun committed:
fix a crash when navigating from UserProfileScreen to ImageDetailScreen

kitakkun committed:
Merge pull request #46 from kitakkun/feature/fix-crashes-when-navigating

fix a crash when navigating from UserProfileScreen to ImageDetailScreen

kitakkun committed:
fix onContentClick events

kitakkun committed:
Merge pull request #47 from kitakkun/feature/navigate-to-post-on-media-tab

fix onContentClick events

kitakkun committed:
add labels on the bottom navigation bar

kitakkun committed:
Merge pull request #48 from kitakkun/feature/improve-bottom-nav-design

add labels on the bottom navigation bar

kitakkun committed:
implement attachments removal

kitakkun committed:
add AnimatedRemovableAttachment

kitakkun committed:
Merge pull request #49 from kitakkun/feature/removable-attachments-on-post-create

Feature/removable attachments on post create

kitakkun committed:
add loading indicator on HomeScreen

kitakkun committed:
update auto fetch timing on UserProfileScreen

kitakkun committed:
Merge pull request #50 from kitakkun/feature/fix-auto-fetch-on-userprofilescreen

update auto fetch timing on UserProfileScreen

kitakkun committed:
add loading indicator on UserProfileScreen

kitakkun committed:
add loading indicator in ReactionsScreen (also enabled auto fetch)

kitakkun committed:
Merge branch 'develop' into feature/loading-indicator

kitakkun committed:
Merge pull request #51 from kitakkun/feature/loading-indicator

Feature/loading indicator

kitakkun committed:
final adjustment

akihiro-n commented on FooS/app/src/main/java/com/example/foos/ui/view/component/list/PostItem.kt:
```diff
@@ -0,0 +1,184 @@
+    modifier: Modifier = Modifier,
+) {
+    createdAt?.let {
+        val createdAtDateTime =
+            createdAt.toInstant().atZone(ZoneId.systemDefault()).toLocalDateTime()
+        val nowDateTime = LocalDateTime.now()
+        val daysDiff = ChronoUnit.DAYS.between(createdAtDateTime, nowDateTime)
+        val hoursDiff = ChronoUnit.HOURS.between(createdAtDateTime, nowDateTime)
+        val minutesDiff = ChronoUnit.MINUTES.between(createdAtDateTime, nowDateTime)
+        val secondsDiff = ChronoUnit.SECONDS.between(createdAtDateTime, nowDateTime)
+        val text: String = if (daysDiff >= 7) {
+            // ex) 9 Jan
+            createdAtDateTime.format(DateTimeFormatter.ofPattern("d MMM"))
+        } else if (daysDiff >= 1) {
+            // ex) 5d
+            "${daysDiff}d"
+        } else if (hoursDiff >= 1) {
+            // ex) 7h
+            "${hoursDiff}h"
+        } else if (minutesDiff >= 1) {
+            // ex) 10m
+            "${minutesDiff}m"
+        } else {
+            // ex) 40s
+            "${secondsDiff}s"
+        }
```
[IMO]この辺りのロジックはViewから切り離したいですね

kitakkun replied:
確かにそうですね！ありがとうございます。
切り離すとすればUseCaseでしょうか？ 🤔

akihiro-n replied:
まず、ここで作られている`text`は最終的にViewで表示するための値なのでUIStateに定義すべきかなと思っています。
その上でロジック自体は「ドメイン層から受け取った値をUIStateに変換している箇所」に記載するのが良いかなと。
現状の実装だとViewModelですかね。あとはUIStateにセカンダリコンストラクタ定義して、そこに記載するみたいな方法でも良いかもしれません。

kitakkun replied:
ありがとうございます！このような表示ロジックを持たせる必要のあるスクリーンが2つありまして、HomeScreenViewModelとUserProfileViewModelがあります。両者に同じコードを記述するのを回避するにはどうするのが適切でしょうか。
また、再描画毎に表示の更新をしたいのですが、これは例えばこのComposableにonRecomposeなどラムダ引数を持たせ、再コンポーズ時にViewModelの関数がトリガーされる形で実装するのが適切でしょうか。（ 例： 1s と表示されていたものが、画面外に行き、戻ってきたとき正しい時間 12s に更新される）

kitakkun replied:
個人的には専用のUseCaseを作ってそれぞれのビューモデルから呼び出しを行うのがいいかなと思っています。あとは、PostItemUiStateに変数が余分に増えてしまうのは避けたいなというのもあります。表示上はcreatedAtのDate型の値が存在すれば問題ないですし、投稿詳細画面では通常の日時表示を行いたいので、悩みどころです。

akihiro-n replied:
> 個人的には専用のUseCaseを作ってそれぞれのビューモデルから呼び出しを行うのがいいかなと思っています

そうですね。
UseCaseも色々とあると思うので一概には言えないところですが
このアプリの設計から考えるとUseCaseの役割はあくまでデータ層からの複数のデータを扱いやすいオブジェクトの単位にまとめるということになると思います。
なので日時の表示形式というUIに直接関わるロジックはUseCaseの責務を超えている気もしますし、それを認めることでUseCaseがなんでもありになっていく可能性が高そうだなと。

個人的には今回のように日時表示のような特に汎用的な機能に関しては、特別にHelperクラスなどを用意してあえてアーキテクチャから外れたイレギュラーな場所で管理するのもアリかなと思っています。

ただ、ここは人によって意見が別れそうなところです。

> PostItemUiStateに変数が余分に増えてしまうのは避けたいなというのもあります

管理する状態を増やしたくないという気持ちはわかります。
ただ、変数として定義しておくことでUiStateを見ればその画面で何を表示しているのか一目でわかるメリットもあります。

あと、どうしても状態を増やしたく無いのであれば下記のような実装もあります。
表示する値を関数で定義することで変数を減らし、副作用の起こりにくいコードにはなります。
```
class HogeUiState(
   val date: Date
) {
   fun dateText(): String {
      // dateから表示用文字列に変換して返す
   }
}
```

akihiro-n replied:
https://developer.android.com/jetpack/guide/domain-layer?hl=ja#common-tasks
GoogleのアーキテクチャのドキュメントにはUIで繰り返し使う`ビジネスロジック`に関してはUseCaseにすべきだとも書いてあるので、上手く切り分けてUseCase側に一部の処理を持っていくのはアリかもしれません。

kitakkun replied:
- UseCaseが何でもアリになっていき後に破綻する恐れがあること
- 管理する状態を増やしたくないこと

これらを考慮して、最終的にPostItemUiState内部にフォーマットされた作成日時を取得する関数を追加することで対応しました。改めまして、丁寧なレビューありがとうございます 🙇 

https://github.com/kitakkun/FooS/pull/60/files

akihiro-n commented on FooS/app/src/main/java/com/example/foos/data/repository/PostsRepository.kt:
```diff
@@ -0,0 +1,226 @@
+        val document = Firebase.firestore.collection(COLLECTION).document()
+        val imageDownloadLinks = mutableListOf<String>()
+        // ファイルの圧縮
+        var i = 1
+        databasePost.attachedImages.forEach {
+            val bitmap = BitmapFactory.decodeFile(it.removePrefix("file://"))
+            val resized = ImageConverter.resize(bitmap, MAX_UPLOAD_IMAGE_SIZE, true)
```
[IMO]`forEachIndexed()`を使えば`i`をインクリメントしなくても良さそう？

kitakkun replied:
修正しました！

kitakkun committed:
refactor 'create' method on PostsRepository

replace forEach with forEachIndexed, and move variable 'i' inside forEachIndexed block

morizooo commented on FooS/app/build.gradle:
```diff
@@ -26,17 +31,21 @@ android {
     compileOptions {
+        sourceCompatibility JavaVersion.VERSION_11
+        targetCompatibility JavaVersion.VERSION_11
         sourceCompatibility JavaVersion.VERSION_1_8
```
[Q]VERSION_1_8の部分必要なのかな？

kitakkun replied:
不要でした！削除しました
https://github.com/kitakkun/FooS/pull/52/commits/0468aaf213a4e66589e4747c55bf9cedafbb3177

morizooo commented on FooS/app/src/main/java/com/example/foos/FirebaseAuthManager.kt:
```diff
@@ -0,0 +1,26 @@
+object FirebaseAuthManager {
+    private var auth: FirebaseAuth = Firebase.auth
+
+    fun checkSignInState(context: Context) {
```
[Q] checkといメソッドなのに副作用発生する？

morizooo commented on FooS/app/src/main/java/com/example/foos/FirebaseAuthManager.kt:
```diff
@@ -0,0 +1,26 @@
+            context.startActivity(
+                Intent(context, SignInActivity::class.java)
+            )
+            val activity = context as? Activity
```
[MUST] これは凶悪なコードなので辞めたほうがいいです。意図せず画面が閉じられてしまう場合があります。必要があるなら呼び出し側で閉じるべきです

morizooo commented on FooS/app/src/main/java/com/example/foos/SignInActivity.kt:
```diff
@@ -0,0 +1,110 @@
+        private const val TAG = "SIGN_IN_ACTIVITY"
+    }
+
+    private lateinit var auth: FirebaseAuth
```
[nits] auth使ってないような

morizooo commented on FooS/app/src/main/java/com/example/foos/data/domain/converter/uistate/ConvertPostToUiStateUseCase.kt:
```diff
@@ -0,0 +1,24 @@
+
+class ConvertPostToUiStateUseCase {
+
+    operator fun invoke(post: Post) : PostItemUiState {
```
[IMO] 口頭で話した思うけどPostItemUiState側にメソッドはやしてConvert系のUseCaseは消す

kitakkun replied:
UiStateへの変換UseCaseを，UiStateの方にconvertメソッドとして定義し，Convert系UseCaseを削除する対応を行いました！ https://github.com/kitakkun/FooS/pull/54

morizooo commented on FooS/app/src/main/java/com/example/foos/data/domain/fetcher/post/FetchPostByDatabasePostUseCase.kt:
```diff
@@ -0,0 +1,45 @@
+        var reactions: List<DatabaseReaction> = listOf()
+
+        coroutineScope {
+            jobs.add(async {
```
[IMO] 全体的にjobsを定義して利用しているが、不要なところでも使っているようにみえるしasync内のエラーの制御もしてないので怪しいです。

joinAllを使うのは後回しにして、他の最適化をして問題があってから使う流れでいいと思います

morizooo commented on FooS/app/src/main/java/com/example/foos/data/repository/FollowRepository.kt:
```diff
@@ -0,0 +1,94 @@
+/**
+ * フォロー状態を管理するリポジトリ
+ */
+object FollowRepository {
```
[MUST] DIで使っているのでinterfaceで定義しImplを実装するのが良いかと

kitakkun replied:
https://github.com/kitakkun/FooS/pull/53
こちらのプルリクエストで全てのリポジトリに関して，インタフェース定義と実装部へ分割を行いました！
Firebase周りのDIは別のfeatureブランチ(https://github.com/kitakkun/FooS/tree/feature/inject-firebase-references )で作業中です．

morizooo commented on FooS/app/src/main/java/com/example/foos/data/repository/FollowRepository.kt:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    private suspend fun isFollowing(from: String, to: String): Boolean {
+        return !Firebase.firestore.collection(COLLECTION)
```
[IMO] FirebaseFirestoreをModuleに追加してコンストラクタで受け取るようにしていいかなと

kitakkun replied:
https://github.com/kitakkun/FooS/pull/55
こちらのプルリクで対応しました！

morizooo commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/userprofile/UserProfileScreen.kt:
```diff
@@ -0,0 +1,321 @@
+}
+
+@Composable
+fun Biography(
```
[nits] これわざわざ定義する必要合ったのかな？

morizooo commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/userprofile/UserProfileScreen.kt:
```diff
@@ -0,0 +1,321 @@
+}
+
+@Preview(showBackground = true, showSystemUi = true)
+@Composable
```
[IMO] UserProfileだけじゃなくて作ったComposable単位でPreviewが合ったほうがいいと思う
Composable単位でファイルわけたほうがいいかな(分けるほどでも無いと思うものはわざわざViewを切り出す必要が無い気がする

morizooo commented on FooS/app/src/main/java/com/example/foos/util/FIleUtils.kt:
```diff
@@ -0,0 +1,207 @@
+    fun getRealPath(context: Context, fileUri: Uri): String? {
+        val realPath: String?
+        // SDK < API11
+        realPath = if (Build.VERSION.SDK_INT < 11) {
```
[IMO] コピペしたと思うけど不要な分岐入ってる。参照元のリンクを書いたほうがいい

morizooo reviewed (commented):
おおまかにしか見てないけどまずは以下の点を見直してほしいかな
・全体的にjobs.joinAll()を使わないようにしてほしい
・DIで使っているはずのにobjectで実態に依存しているののでinterfaceで依存するようにしてほしい
・ローカルで動かずpreviewもみれないので、誰でも使えるようにしてほしい

morizooo commented on FooS/app/src/main/java/com/example/foos/ui/view/component/ConfirmDialog.kt:
```diff
@@ -0,0 +1,32 @@
+import androidx.compose.material.TextButton
+import androidx.compose.runtime.Composable
+
+@Composable
```
[IMO] 基本的にViewを定義したらPreviewを書いたほうがいいです

morizooo commented on FooS/app/src/main/java/com/example/foos/ui/view/component/FollowButton.kt:
```diff
@@ -0,0 +1,20 @@
+import com.example.foos.R
+
+@Composable
+fun FollowButton(
```
[IMO] 全体的にVIewを作ったならにPreviewを書いたほうがいいです

kitakkun committed:
add files to build and execute the app

kitakkun committed:
replace MAPS_API_KEY with a placeholder

ryu38 commented on FooS/app/src/main/java/com/example/foos/SignInActivity.kt:
```diff
@@ -0,0 +1,110 @@
+            Log.d(TAG, "Sign in successful!")
+            goToMainActivity()
+        } else {
+            Toast.makeText(this, "There was an error signing in", Toast.LENGTH_LONG).show()
```
[IMO]アプリ内で表示する文言は文字列リソースから取るようにしよう

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/MainScreen.kt:
```diff
@@ -0,0 +1,142 @@
+fun MainScreen() {
+
+    // 認証済みか確認し、未認証であれば認証を行う
+    FirebaseAuthManager.checkSignInState(LocalContext.current)
```
[IMO]この処理によって起こる変化がぱっと見わからないから、ViewModelでイベントとして処理した方が良さそう

ryu38 replied:
下のScreenViewModelで処理するのはどう？

kitakkun replied:
色々試行錯誤したのですが、MainActivity内のonStart内で実装することに一旦落ち着きました。結局のところ認証していないとアプリケーションの機能は使えないのでMainActivityで認証してしまうのが直感的だし実装も容易かなという判断です。

ScreenViewModelで認証処理を実装するのも試みたのですが、`registerForActivityResult`がコールできないなど、制約が多く難しかったです。ちなみに公式ドキュメントでもActivity内でやっているようです（https://firebase.google.com/docs/auth/android/start?hl=ja ）

該当の変更内容を含むコミットはこちらです↓
https://github.com/kitakkun/FooS/pull/52/commits/fe47bfeb03bb21e6bf06d04456121c423f091587

お手数ですが確認いただけますと幸いです 🙏

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/MainScreen.kt:
```diff
@@ -0,0 +1,142 @@
+fun RowScope.MyBottomNavigationItem(
+    screen: Screen,
+    currentDestination: NavDestination?,
+    navController: NavHostController,
```
[nits]navController使ってないよ！

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/MainScreen.kt:
```diff
@@ -0,0 +1,142 @@
+    )
+    val navBackStackEntry by navController.currentBackStackEntryAsState()
+    val currentDestination = navBackStackEntry?.destination
+    if (screens.map { it.route }.contains(currentDestination?.route)) {
```
[Q]ナビゲーションが出現するかしないかの処理はここでやるべきでない気がするがどうだろう

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/home/HomeScreen.kt:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    // Bottomナビゲーションでホームがクリックされたらトップへスクロール
+    LaunchedEffect(Unit) {
```
[IMO]こんな感じに一つにまとめられます！
```
LaunchedEffect(Unit) {
    launch {
        viewModel.navEvent.collect { ... }
    }
    launch {
        viewModel.navRoute.collect { ... }
    }
}
```

kitakkun replied:
知りませんでした！コードが読みにくくなって困っていたところでした．ありがとうございます 🙇

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/home/HomeScreen.kt:
```diff
@@ -0,0 +1,75 @@
+    val listState = rememberLazyListState()
+
+    // 初回起動時の投稿フェッチ
+    LaunchedEffect(uiState.posts.isEmpty()) {
```
[IMO]これ初回だけならUnitでいい気がする

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/MainScreen.kt:
```diff
@@ -0,0 +1,142 @@
+ * 画面下部のナビゲーションバー
+ */
+@Composable
+fun ScreenBottomNavBar(
```
[IMO]ナビゲーションバーは別ファイルにおいた方がいい気がす

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/MainScreen.kt:
```diff
@@ -0,0 +1,142 @@
+ * 画面トップナビゲーションバー
+ */
+@Composable
+fun ScreenTopBar(
```
[IMO]これも別ファイル

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/home/HomeScreen.kt:
```diff
@@ -0,0 +1,75 @@
+        state = rememberSwipeRefreshState(isRefreshing = uiState.isRefreshing),
+        onRefresh = { viewModel.onRefresh() }
+    ) {
+        if (uiState.posts.isEmpty()) {
```
[IMO]投稿がゼロ==読み込み中ではないはずなので、`isLoading`的な状態をつくろう

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/followlist/FollowListScreen.kt:
```diff
@@ -0,0 +1,150 @@
+    )
+
+    val pagerState = rememberPagerState(initialPage)
+    val coroutineScope = rememberCoroutineScope()
```
[nits]使ってないよ！

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/followlist/FollowListScreen.kt:
```diff
@@ -0,0 +1,150 @@
+) {
+    SideEffect {
+        fetchEvent()
+    }
```
[Q]ここ初回の読み込み処理ならばSideEffectじゃなくてLaunchedEffect(Unit)であるべきですけどどうですか？

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/followlist/FollowListScreen.kt:
```diff
@@ -0,0 +1,150 @@
+}
+
+@Composable
+fun FollowerList(
```
[IMO]FolloweeListと同じなのでどっちかいらないかと

ryu38 replied:
今のUserListを解体してFollowerList(もしくはFolloweeList)をUserListと名付けた方がいいかな

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/component/list/UserList.kt:
```diff
@@ -0,0 +1,29 @@
+import com.example.foos.ui.view.component.OnAppearLastItem
+
+@Composable
+fun UserList(
```
[IMO]これ中身LazyColumnラップしてるだけなので解体でいいかも

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/followlist/FollowListViewModel.kt:
```diff
@@ -0,0 +1,83 @@
+@HiltViewModel
+class FollowListViewModel @Inject constructor(
+    private val usersRepository: UsersRepository,
+    private val followRepository: FollowRepository,
```
[nits]使ってないぞ〜

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/component/list/ReactionList.kt:
```diff
@@ -0,0 +1,27 @@
+import com.example.foos.ui.state.screen.reaction.ReactionItemUiState
+
+@Composable
+fun ReactionItemList(
```
[IMO]作ってる途中感あるけど、LazyColumnをラップするだけなら冗長かな

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/component/extensions/LazyListStateExtensions.kt:
```diff
@@ -0,0 +1,23 @@
+package com.example.foos.ui.view.component
```
[MUST]パッケージとファイルの場所が一致してないので直そう

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/post/PostScreen.kt:
```diff
@@ -0,0 +1,240 @@
+    Log.d("LOCATION", locationAttached.toString())
+
+    if (locationAttached) {
+        viewModel.applyLocationData(location!!, locationName!!)
```
[IMO]強制アンラップはなるべく避けたいんじゃ〜

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/view/screen/post/PostScreen.kt:
```diff
@@ -0,0 +1,240 @@
+    )
+    {
+        viewModel.setImages(context, it)
+    }
```
[nits]書き方が変になってます

kanesy commented on FooS/app/src/main/java/com/example/foos/MainActivity.kt:
```diff
@@ -3,41 +3,23 @@ package com.example.foos
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+
+        val config = BundledEmojiCompatConfig(applicationContext)
+        EmojiCompat.init(config)
```
[nits]
めっちゃ細かいけど、こういうのはApplication側で処理したほうが良いですよ。
（[公式の例](https://developer.android.com/guide/topics/ui/look-and-feel/emoji-compat?hl=ja#downloadable-fonts-config)もApplication.onCreate）

Application.onCreateはアプリのエントリポイント
XXActivity.onCreateはXX画面のエントリポイント

EmojiCompat.init()は、どのスコープで作用する処理？画面？アプリ？
と考えると、現状の実装に違和感が出ると思う

kitakkun replied:
こちらのコミット  https://github.com/kitakkun/FooS/pull/52/commits/19305e2260c6195bc29c0af1c3e40cb523e8057b にて修正を行いました！細かいことですが，とても重要なことですね！確かにこの記述では，Activityが自身の役割を超えた処理を行なっていることになると思います．

kitakkun committed:
separate PostsRepository into its interface and implementation

ryu38 commented on FooS/app/src/main/java/com/example/foos/ui/state/screen/postdetail/PostDetailScreenUiState.kt:
```diff
@@ -0,0 +1,7 @@
+
+data class PostDetailScreenUiState(
+    val postItemUiState: PostItemUiState
+)
```
[IMO]PostItemUiStateしかないならいらないと思う

kitakkun committed:
separate FollowRepository into its interface and implementation

kanesy commented on FooS/app/src/main/java/com/example/foos/SignInActivity.kt:
```diff
@@ -0,0 +1,110 @@
+        super.onStart()
+
+        if (Firebase.auth.currentUser == null) {
+            val signInIntent = AuthUI.getInstance()
```
[感想]
AuthUIなんてあるんだ。便利...。
（いつもFirebaseAuth使って自前で画面作ってた）

kitakkun replied:
事前に用意されているAuthUIに頼ってめんどくさい部分を全投げした感じなので，どちらかと言うと逆に自前で好きな見た目の認証UIを作成する方法を知りたいですね．．． 🤔  色々調べれば多少なり出てくるとは思いますが大変そうです

kitakkun committed:
separate ReactionsRepository into its interface and implementation

kitakkun committed:
separate UsersRepository into its interface and implementation

kitakkun committed:
Merge pull request #53 from kitakkun/feature/separate-repository-into-interface-and-implementation

Feature/separate repository into interface and implementation

kitakkun committed:
Create FirebaseModule.kt

kitakkun committed:
Inject Firebase dependencies on FollowRepository

kitakkun committed:
Inject Firebase dependency on PostsRepository

kitakkun committed:
Inject Firebase dependency on ReactionsRepository

kitakkun committed:
Inject Firebase dependency on UsersRepository

kitakkun committed:
redefine ConvertPostToUiStateUseCase as a member static function of PostItemUiState

kitakkun committed:
redefine ConvertReactionToUiStateUseCase as a member static function of ReactionItemUiState

kitakkun committed:
Merge pull request #54 from kitakkun/feature/move-converter-usecases-into-ui-states

Feature/move converter usecases into UI states

kitakkun committed:
move emoji initialization into MyApplication.kt

kitakkun committed:
remove FirebaseStorage

kitakkun committed:
Merge pull request #55 from kitakkun/feature/inject-firebase-references

Feature/inject firebase references

kitakkun committed:
add preview of ConfirmAlertDialog

kitakkun committed:
add FollowButtonPreview

kitakkun committed:
make ConfirmAlertDialogPreview private

kitakkun committed:
add UserIconPreview

kitakkun committed:
add MaxSizeLoadingIndicatorPreview

kitakkun committed:
replace async and jobs.joinAll logics with procedural logics

kitakkun committed:
refactor PostsRepository and FollowRepository

kitakkun committed:
remove chunk logic from usecases

kitakkun committed:
add chunk logic on ReactionsRepository

kitakkun committed:
Merge pull request #56 from kitakkun/feature/remove-unnecessary-async

Feature/remove unnecessary async

kitakkun committed:
fix infinity loading indicator on HomeScreen and UserProfileScreen

kitakkun committed:
Merge pull request #57 from kitakkun/feature/fix-infinity-loading

fix infinity loading indicator on HomeScreen and UserProfileScreen

kitakkun committed:
fix whereIn query crashes on repositories when providing empty list as an argument

kitakkun committed:
Merge pull request #58 from kitakkun/feature/fix-where-in-query-crash-on-giving-empty-list

fix whereIn query crashes on repositories when providing empty list

kitakkun committed:
remove unnecessary compile options

kitakkun committed:
fix wrong package name in LazyListStateExtensions.kt

kitakkun committed:
separate MenuItemList and MenuItem into different files

kitakkun committed:
move Button components into button package

kitakkun committed:
add preview of RoundIconActionButton

kitakkun committed:
theme Previews

kitakkun committed:
remove unnecessary composable functions

kitakkun committed:
extract ReactionButton and ReactionDropdown

kitakkun committed:
move ConfirmDialog into dialog package

kitakkun committed:
recreate sign-in and sign-out feature

delete SignInActivity. place code that launches Firebase AuthUI into MainActivity.

kitakkun committed:
extract string resources

kitakkun committed:
Merge pull request #59 from kitakkun/feature/refactor-auth-process

refactor Firebase authorization process.

kitakkun committed:
fix TAG when logging

kitakkun committed:
extract date format logic into PostItemUiState

kitakkun committed:
Merge pull request #60 from kitakkun/feature/extract-date-format-logic

extract date format logic into PostItemUiState

kitakkun committed:
extract FollowInfo into a file

kitakkun committed:
remove unnecessary code in SettingScreen, fix result handling of CropImageView

kitakkun committed:
fix weird format of code

kitakkun committed:
add Preview of FollowListScreen

kitakkun committed:
create PostUI Preview

kitakkun committed:
create HomeUI Preview

kitakkun committed:
add isLoadingPosts parameter to HomeScreenUiState

kitakkun committed:
create SettingUIPreview

kitakkun committed:
remove unnecessary line break

kitakkun committed:
create LocationConfirmUIPreview

kitakkun committed:
disable map interaction on LocationConfirmUI

kitakkun committed:
create LocationSelectUIPreview

kitakkun committed:
disable confirm button when location is null

kitakkun committed:
create ImageDetailUIPreview, also add TopAppBar

kitakkun committed:
fix TopAppBar behind contents

kitakkun committed:
adjust TopAppBar style

kitakkun committed:
fix a problem that backArrow icon is difficult to see

kitakkun committed:
create MapUIPreview

kitakkun committed:
create ReactionUIPreview

kitakkun committed:
create attachment previews

kitakkun committed:
create PostDetailPreview

kitakkun committed:
refactor FollowInfo composable

kitakkun committed:
create RoundButtonPreview

kitakkun committed:
refactor RoundIconActionButton

kitakkun committed:
Merge branch 'develop' into feature/write-previews-of-composables

kitakkun committed:
Merge pull request #61 from kitakkun/feature/write-previews-of-composables

Feature/write previews of composables
